### ISSUE TITLE
Fully integrate XBRL based FERC filings

### ISSUE THREAD

**Issue by zschira** [catalyst-cooperative/pudl#1574]
# Background
[This](https://github.com/catalyst-cooperative/pudl/issues/1568) Epic focuses on integrating XBRL based FERC tables that are currently in use by PUDL or by RMI. Once this is complete we will want to make all XBRL based data compatible with historical data. This means we should be able to seamlessly extract and transform XBRL filings to the structure of the raw FERC Form 1 database. The finished product for this work should likely look like an updated version of the `ferc1_to_sqlite` script that can be given a set of years including those with only XBRL data, and those with only historical data available, and produce a database with all tables for all years.

**Comment by zaneselvans** [comment]
Given the need to integrate a whole second stream of data extraction and transformation into the FERC 1 process here, and the fact that we have wanted to disentangle some of the extraction vs. transformation code (deferring some complicated steps until later in the process) it might be worth thinking about those tasks together. Rather than shoehorning all of the XBRL data into the existing pipeline, and then subsequently disassembling the now even more complicated pipeline.

**Comment by zaneselvans** [comment]
It seems like after the data is in the PUDL DB, there will be another layer of work that has to happen to integrate it with the FERC-EIA record linkage (and any other downstream analysis), since some of the old columns (`record_id`) won't exist any more.

**Comment by cmgosnell** [comment]
I have lots of thoughts but will respond quickly to this:
>  Should it have its own Zenodo archive, or should it be combined with the existing FERC Form 1 archive?

I think it should have its own archive! They will lead into different "pre-extract" steps (archive -> raw SQLite)

**Comment by zaneselvans** [comment]
There are 4 combinations of data source / existing code. In order we want to tackle them:
* PUDL tables from XBRL
* Hub tables from XBRL
* Hub tables from FoxPro
* PUDL tables from FoxPro (into our new transform architecture)

**Comment by cmgosnell** [comment]
Transform step design thoughts:
* we make a generic table transformer w/ some `execute()` or coordinating method that has standard cleaning/renaming/reshaping/encode/etc steps.
* Make clean, nice lil table-specific transformers inheriting from the generic transformer for each of the new XBRL tables
* Make a lil shell of the generic transformer for the old foxpro tables that is mostly a wrapper for the existing transform functions
* Make a concatenater

**Comment by bendnorman** [comment]
Comments on Dagster:

I don't it is possible for part of our ETL to be implemented using Dagster. If the new FERC data was going to be an entirely separate pipeline then I'd say we should use dagster but given it will be intertwined with our existing ETL I don't think we should

However, we should keep some dagster concepts in mind while designing these FERC transformations. I would highly recommend reading the [dagster concepts docs](https://docs.dagster.io/concepts) and our [rough plan](https://docs.google.com/document/d/1_DbHu-cSb33cMph8CN4-YUYT4j_sO4Dz7Xx6S34o1y4/edit#heading=h.9bv65wjtabl7) for applying these concepts to PUDL. 

Each transform function will eventually be turned into a dagster "op" which is a node in the DAG. An op takes some upstream inputs, does some processing then passes the result to the next op in the DAG. Dagster is strict about what can be fed into ops:

> An op only starts to execute once all of its inputs have been resolved. Inputs can be resolved in two ways:
> * The upstream output that the input depends on has been successfully emitted and stored.
> * The input was stubbed through config.

In other words, the input of an op has to be the output of another op OR configuration. Configuration can come in the form of [Resources](https://docs.dagster.io/concepts/resources) and [Config](https://docs.dagster.io/concepts/configuration/config-schema) which allow you to configure DAG wide resources in a yaml file. ETL wide abstractions like the Datastore, pudl_settings and dataset settings will likely become Resources or Config. This means ops won't accept python constants passed into functions.  For example:

```python
@op
def extract():
      return pd.read_csv("poopy_data.csv")

@op
def fillnans(df, fill_value):
      return df.fillna(fill_value)

@graph
def run_etl()
      df = extract()
      fillnans(df, "Gas")
```
won't work because the `fill_value` parameter is a python string, not the output of another op or a Dagster Resource or Config. You could fix this by configuring the `fill_value` parameter using Dagster Configs. 

Also, these rules input rules only apply to functions you think should be included in the DAG. For example, helper functions probably won't be included in the DAG. 

```python
def fillnans(df, fill_value):
      return df.fillna(fill_value)

@op
def transform(df):
     df = fillnans(df, "Gas")
     return df
```
 
The dagster refactor and xbrl stuff will probably be happening in parallel so Zane and I will probably have more guidance as we refactor!

**Comment by bendnorman** [comment]
Some more thoughts on Dagster:

Dagster allows for created nested graphs to allow for some logic hierarchy. @zaneselvans and I envision incrementally applying dagster graphs to our ETL. There are multiple levels we've identified:

1. Turn our `_etl_{dataset}()` functions into ops and construct a graph. This is a very simple DAG but would enable us to run the ETLs in separate processes. 
2. Turn our ETL functions into graphs where the ops are the extract and transform steps.
3. ETL functions are graphs, E and T steps are graphs and individual table transforms are ops.
4. ETL functions are graphs, E and T steps are graphs, individual table transforms are graphs and individual reusable cleaning functions are ops. 

When we're writing XBRL functions we don't need to think much about the first two levels because we are only making changes within the `ferc1.transform()` function. 

## Option 3
I wrote up a [little prototype](https://github.com/catalyst-cooperative/pudl/blob/aa36c91766e70eb3a178232655f7c79fb60b6a87/notebooks/work-in-progress/ferc_dagster_prototype.py) for option 3. It seems pretty clean but it seems awkward to have to define all of the transformed tables in the transform graph out:

https://github.com/catalyst-cooperative/pudl/blob/aa36c91766e70eb3a178232655f7c79fb60b6a87/notebooks/work-in-progress/ferc_dagster_prototype.py#L25-L33

Satisfying option three shouldn't be that difficult because we can use the existing structure in `transform.ferc1`. I do have a couple of questions:
- Do the table transform functions depend on previously cleaned tables? It looks like the `plants_steam()` depends on the `fuel_ferc1()` table for plant_id assignment. 
- Do the transform functions depend on multiple tables in `ferc1_raw_dfs`?

## Option 4
Option 4 is a bit trickier because we want generic cleaning functions that can be parameterized for each table but not have to use dagster config. [Op factories](https://docs.dagster.io/concepts/ops-jobs-graphs/ops#op-factory) are the solution here but I haven't been able to get them working yet.

**Comment by zschira** [comment]
This is kind of an out there idea, but could we develop our own op factory decorator that makes creating op factories a little less verbose. Something like:

```py
from functools import partial

def op_factory(name="default_name", ins=None, **kwargs):
    def decorator(op_func):
        @wraps(op_func)
        def wrapper(**op_kwargs):
            return op(name=name, ins=ins, **kwargs)(partial(op_func, **op_kwargs))
```

I'm not 100% sure that this would work as is (decorators are confusing), but the idea here is that it could be applied like:

```py
@op_factory
def rename_df_factory(context, df, column_mapping=None):
    context.log.info(f"\n The DataFrame: {df}\n")
    context.log.info(f"\n The Op Ins: {context.op_def.ins}\n")
    t_df = df.rename(columns=column_mapping)
    context.log.info(f"\n The Transformed DataFrame: {t_df}\n")
    return t_df

@op
def extract():
    return pd.DataFrame([1,2], columns=["col"])

@job()
def etl():
    df = extract()
    column_mapping = {"col": "column"}
    transformed_df = rename_df_factory(column_mapping=column_mapping)(df)
```

Here, when you call `rename_df_factory` with `column_mapping` as a keyword arg, the decorator will create a partially specified method, then apply the `op` decorator to that method, and return the created `op`.

**Comment by bendnorman** [comment]
## Vocab:
- **Table transform functions** are responsible for cleaning one table form a datasource. The functions typically accept some raw data and return a cleaned dataframe. 
- **Generic cleaning functions** are specific steps within table transform functions. For example, renaming columns, filling na values, backfills... These functions typically accept a dataframe to clean and some parameters to specify how the data should be cleaned. For example, a column rename dictionary, fillna value...

## Background
We want to develop a way to convert generic cleaning functions to dagster ops so they are documented in the DAG and could be validated using [dagster-pandera](https://docs.dagster.io/integrations/pandera). 

There are two ways to configure ops, [`ConfigSchema`](https://docs.dagster.io/concepts/configuration/config-schema) and [op factories](https://docs.dagster.io/concepts/ops-jobs-graphs/ops#op-factory). Here are a couple of options for Dagster-friendly generic transform functions:

## Option 1: Use Dagster `ConfigSchema`s for each generic transform function
Description of `ConfigSchema` from dagster docs:
> Various Dagster entities (ops, assets, resources) can be individually configured. When launching a job that executes (ops), materializes (assets), or instantiates (resources) a configurable entity, you can provide run configuration for each entity. Within the function that defines the entity, you can access the passed-in configuration off of the context. Typically, the provided run configuration values correspond to a configuration schema attached to the op/asset/resource definition. Dagster validates the run configuration against the schema and proceeds only if validation is successful.

> A common use of configuration is for a [schedule](https://docs.dagster.io/concepts/partitions-schedules-sensors/schedules) or [sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/schedules) to provide configuration to the job run it is launching. For example, a daily schedule might provide the day it's running on to one of the ops as a config value, and that op might use that config value to decide what day's data to read.

Configs are mostly used for parameters that are changed between runs. For example, a `pudl_out` path or dataset processing settings. We initially did not entertain this option because these generic cleaning functions will be used dozens of times and the parameters will not change for different runs. Given we need to store all of the transformation parameters somewhere, we realized it might make sense to set a default dagster config with all of the static generic transform function metadata. 

You can view the pseudo code [here](https://github.com/catalyst-cooperative/pudl/blob/xbrl-dagster-prototype/notebooks/work-in-progress/xbrl_dagster_prototypes/config_option.py). 

### Pros
 - Can overwrite transform parameters at runtime. (Is this desirable?)
 - Don't have to use dagster factories, which might confuse some developers.

### Cons
- Dagster doesn't allow for duplicate op names. For example, the `transform_steam` graph has an op named `rename_columns` and an op named `rename_columns_2` because `rename_columns()` is called for the dbf and xbrl dataframes. This prevents us from understanding what config is associated with xbrl and dbf.
- Converting our transform metadata to a dagster config might be tricky.
- The generic cleaning functions might be hard to test because the parameters come from the dagster `context` keyword. 

## Option 2: Pass parameters from transform metadata classes into op factories
This option grabs the transform metadata for a given table in a table transform function and passes the relevant data to an op factory of a generic cleaning function. 

You can view the pseudo-code [here](https://github.com/catalyst-cooperative/pudl/blob/xbrl-dagster-prototype/notebooks/work-in-progress/xbrl_dagster_prototypes/op_factory_option.py). 

### Pros
- Easier to test the generic cleaning functions because the `context` keyword isn't needed. 
- Can specify the op name instead of having the auto-incrementing suffix for duplicate ops. 

### Cons
- Need to specify the name of the transform op. The pseudo-code manually specifies the name, but it could probably be programmatically generated, given the metadata. 
- Need to create a factory for every cleaning function. We could probably create a decorator that turns a python function into a parameterizable op. 

## Option 3: Lookup relevant transform metadata within op factory
This option grabs the relevant transform parameters from the transform metadata structure given the table name. 

You can view the pseudo-code [here](https://github.com/catalyst-cooperative/pudl/blob/xbrl-dagster-prototype/notebooks/work-in-progress/xbrl_dagster_prototypes/metadata_lookup_option.py). 

### Pros
- Only need to pass the name of the table name to the op factory.

### Cons
- Increasing coupling. The op factories need to know what the metadata classes look like.

## Option 4: Use transformer class instead of dagster ops for generic cleaning functions
We might get diminishing returns applying dagster deeper into our call stack. Are the UI documentation and potential pandera validation worth the complexity of converting generic cleaning functions into ops? Most generic cleaning functions will be applied sequentially so there won't be complex dependencies to manage.

**Comment by cmgosnell** [comment]
@bendnorman I think I understand these options and it seems like option 2 is our best case at least to move things forward before we fully integrate dagster as our orchestrator. Does this sound right to you?

**Comment by bendnorman** [comment]
Yeah, I think option 2 or 4 are the front runners. All options rely on a non-dagster data structure to store cleaning metadata. If we went with option 2, y'all could just create a library of generic cleaning functions that can be converted to ops. If we go with option 4, the cleaning functions would likely be methods in a transformer class that couldn't be easily converted to ops.

**Comment by bendnorman** [comment]
I'm kind of cooling off on converting cleaning functions to dagster ops:
- One reason for converting cleaning functions to dagster ops is to validate data after each cleaning operation using [dagster-pandera](https://docs.dagster.io/integrations/pandera). How important is this to us? I think unit testing cleaning functions and using dagster-pandera to validate the outputs of table transform ops should be adequate. We also haven't talked about how to store and access the hundreds of pandera schemas for each cleaning step of every PUDL table.
- Cleaning functions could be documented in the dagit UI and parallelized if they are ops. A majority of cleaning functions are applied sequentially, so there aren't many complex dependencies or opportunities to parallelize. It would be nice to view all of the cleaning functions applied to a table in Dagit, but users could also just visit the source code. I posted about this design question in the dagster community, and one engineer at dagster thought it might make more sense to keep our cleaning functions as pure python functions or methods.
- It seems like many contributors want to tweak how tables are transformed. If we kept the cleaning functions as pure python, these contributors wouldn't need to learn dagster concepts. However, people contributing new datasets or tables would need to wrap functions in dagster abstractions. 
- Storing transform metadata and logic in classes could reduce coupling. This way, table transform functions, and cleaning functions don't need to know about the structure of the metadata.

**Comment by zaneselvans** [comment]
I'm excited to use the integrated dataframe schemas provided by the Pandera/Dagster integration, but there are going to be a whole lot of table-level transformations and storing schemas for every one of them is probably overkill. I think we're going to be persisting more intermediate steps with the new setup than we were previously, and probably it makes sense to just check the schemas of the dataframes that we're persisting, since we'll need to think about those schemas for the database anyway.

It seems like something analogous to **Option 4** could also be implemented by composing functions rather than using classes to store the transformation logic. The parameterizations of the transformations would be stored in the Pydantic classes, and then they would be consumed as arguments by individual transform functions. Only the top level table-specific transform functions (the functions containing all of the operations to be applied between persisted versions of a given table, like `plants_steam_ferc1`) would be Dagster ops, and they would know what table they pertain to, and be able to call both generic cleaning functions (looking up their parameters in the Pydantic data structure using the table ID) and table-specific functions which could be defined within the table-level cleaning function, or declared as `_private_functions()` elsewhere in the module.

I'm not sure which of these would feel cleaner, and will write some pseudo-code trying them both out.

**Comment by jdangerx** [comment]
@zschira @zaneselvans does this issue still feel like a useful container for tracking not-done FERC work? Seems like we should have more tightly-scoped issues for the individual FERC-related pushes we'll eventually have in the future...

**Comment by zaneselvans** [comment]
I agree that this monster should be retired. If we make sure any remaining issues are tagged with the appropriate labels (ferc1, ferc2, xbrl, dbf, etc.) we can find them and re-organize them into a new FERC/XBRL work plan in the future.
